fetcher/fetcher.go

```diff
@@ -0,0 +1,258 @@
+				f.updateDependencies(t.key)
+				continue
+			}
+
+			// Doesn't exist in cache, check if we were the first to
+			// request the key.
+			f.keyLock.RLock()
+			var firstTx ids.ID
+			if len(f.keysToFetch[t.key].queue) > 0 {
+				firstTx = f.keysToFetch[t.key].queue[0]
+			}
+			f.keyLock.RUnlock()
+			if firstTx != t.id {
+				// Fetch other keys instead
+				continue
+			}
+
+			// Fetch from disk once if key isn't already in cache
+			v, err := f.im.GetValue(t.ctx, []byte(t.key))
+			if errors.Is(err, database.ErrNotFound) {
+				f.updateCache(t.key, nil, false, 0)
+				f.updateDependencies(t.key)
+				continue
+			} else if err != nil {
+				f.stopOnce.Do(func() {
+					f.err = err
+					close(f.stop)
+				})
+				return
+			}
+
+			// We verify that the [NumChunks] is already less than the number
+			// added on the write path, so we don't need to do so again here.
+			numChunks, ok := keys.NumChunks(v)
+			if !ok {
+				f.stopOnce.Do(func() {
+					f.err = ErrInvalidKeyValue
+					close(f.stop)
+				})
+				return
+			}
+
+			f.updateCache(t.key, v, true, numChunks)
+			f.updateDependencies(t.key)
+		case <-f.stop:
+			return
+		}
+	}
+}
+
+// Checks if a key is in the cache.
+func (f *Fetcher) has(k string) bool {
+	f.cacheLock.RLock()
+	defer f.cacheLock.RUnlock()
+
+	_, exists := f.cache[k]
+	return exists
+}
+
+// Puts a key that was fetched from disk into cache
+func (f *Fetcher) updateCache(k string, v []byte, exists bool, chunks uint16) {
+	f.cacheLock.Lock()
+	defer f.cacheLock.Unlock()
+
+	f.cache[k] = &fetchData{v, exists, chunks}
+}
+
+// After fetching a key, decrement the tx count that was waiting on the key.
+// This function is only called by the worker.
+func (f *Fetcher) updateDependencies(k string) {
+	f.keyLock.Lock()
+	defer f.keyLock.Unlock()
+
+	txIDs := f.keysToFetch[k].queue
+	f.txnLock.Lock()
+	for _, id := range txIDs {
+		f.txnsToFetch[id].Done() // Notify all other txs
+	}
+	f.txnLock.Unlock()
+	f.keysToFetch[k].queue = nil
+
+	// When the cache check calls this, it should already be true
+	if !f.keysToFetch[k].fetched {
+		f.keysToFetch[k].fetched = true
+	}
+}
+
+// Lookup enqueues keys for the workers to fetch
+func (f *Fetcher) Lookup(ctx context.Context, txID ids.ID, stateKeys state.Keys) *sync.WaitGroup {
+	wg := &sync.WaitGroup{}
+	f.txnLock.Lock()
+	f.txnsToFetch[txID] = wg
+	f.txnLock.Unlock()
+
+	f.keyLock.Lock()
+	tasks := make([]*task, 0, len(stateKeys))
+	for k := range stateKeys {
+		if _, ok := f.keysToFetch[k]; !ok {
+			f.keysToFetch[k] = &keyData{fetched: false, queue: make([]ids.ID, 0)}
+		}
+		// Don't requeue keys that were already fetched
+		if f.keysToFetch[k].fetched {
+			continue
+		}
+		f.keysToFetch[k].queue = append(f.keysToFetch[k].queue, txID)
+		t := &task{
+			ctx: ctx,
+			id:  txID,
+			key: k,
+		}
+		tasks = append(tasks, t)
+	}
+	f.keyLock.Unlock()
+
+	// Create wg based on number of keys we need to wait on
+	wg.Add(len(tasks))
+
+	for _, t := range tasks {
```

Comment:
You need to ensure the fetcher isn't stopped before doing this (otherwise it will panic for adding to a closed channel)